rfcs/20241015-result-accuracy.md

```diff
@@ -0,0 +1,101 @@
+# [RFC] Add result accuracy to transcendental unary ops
+
+Status: In Review<br/>
+Initial version: 10/15/2024<br/>
+Last updated: 10/15/2024<br/>
+Discussion thread:
+
+## Overview
+
+This RFC proposes adding a new attribute `result_accuracy` to the following
+transcendental unary ops: `exp`, `expm1`, `log`, `logp1`, `logistic` and `tanh`.
+`result_accuracy` allows the user to choose the implementation of these ops
+based on the accuracy they request. The choice of implementation is restricted
+to F32 inputs.
+
+## Background
+
+Transcendental ops can have multiple implementations on a processing unit that
+vary in accuracy and performance. By allowing users to select implementation
+based on accuracy per op, we can offer more tools to weigh tradeoffs between
+accuracy and performance and also ensure more consistent numerical behaviors
+across different devices.
+
+### Caveats on targets
+
+The proposed result_accuracy attribute can only be supported for targets that
+have multiple implementations of the ops. For example, for XLA-CPU, the
+implementations of these ops are dependent on LLVM and each target may have a
+different implementation with a different level of accuracy. Thus, further
+analysis needs to be done to support this feature on CPUs. This analysis could
+be performed while building the compiler, assuming the compiler will only be
+used on a single type of CPU.
+
+## Proposed Specification
+
+### `result_accuracy`
+
+The users can specify the worst case numerical error they can tolerate in terms
+of absolute, relative and ULP (unit in last place) errors. If they don't care
+about the numerical accuracy, they can also choose the implementation using
+`mode`. We propose a new attribute `result_accuracy`. `result_accuracy` can be
+any combination of the following numerical tolerances `atol`, `rtol`, `ulps` or
+an enum of `HIGHEST`, `DEFAULT` or `TOLERANCE`. `TOLERANCE` enum is a default
+placeholder value for `mode` when the numerical tolerances are used. When using
+the numerical tolerances, at least one of atol, rtol or ulps should be
+specified.
```

Comment:
It seems to me that mode is optional with default value `TOLERANCE` which necessitates providing values for atol, rtol and ulps. IMO it seems it might be confusing to the user to call something else as default when we have a DEFAULT enum value. But anyways I do not want the proposal to block on this.  


Are the `DEFAULT` and `HIGHEST` enum value refers to the precision values defined in the existing spec.

If not, How about we 
1. remove the `TOLERANCE` value from the enum
2. The `DEFAULT` behavior is to specify the `atol`, `rtol`, and `ulps`, given that these values must tbe provided when mode != TOLERANCE.

 If yes, I have a quick question: is there a reason we do not use HIGH?

Reply:
Thanks for the comment. `DEFAULT` would return a fast but less accurate implementation and `HIGHEST` would give the most accurate implementation. We didn't use `HIGH` because right now we only have two implementations of the ops. Also, we added `TOLERANCE` because when `atol`, `rtol` and `ulps` are set, `TOLERANCE` will disambiguate it from the `DEFAULT` mode.

Reply:
Thanks for the clarification. 
The proposal lgtm with a minor comment.